File: BOJ/java/14501.java
import java.util.Scanner;

class Main {
    static int[] a;
    static int[] p;
    static int n;
    static int result = 0;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        a = new int[n];
        p = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = sc.nextInt();
            p[i] = sc.nextInt();
        }

        go(0, 0);
        System.out.println(result);
    }

    static void go(int day, int total) {
        if (day == n) {
            result = Math.max(result, total);
            return;
        }
        go(day + 1, total);
        if (day + a[day] <= n) {
            go(day + a[day], total + p[day]);
        }

    }
}

File: BOJ/kotlin/14501.kt
import kotlin.math.max

val a: MutableList<Int> = mutableListOf()
val p: MutableList<Int> = mutableListOf()
var n: Int = 0
var result: Int = 0

fun go(day: Int, total: Int) {
    if (day == n) {
        result = max(result, total)
        return
    }

    go(day + 1, total)
    if (day + a[day] <= n) {
        go(day + a[day], total + p[day])
    }
}

fun main() {
    n = readLine()!!.toInt()
    repeat(n) {
        val (day, price) = readLine()!!.split(" ").map { it.toInt() }
        a.add(day)
        p.add(price)
    }
    go(0, 0)
    println(result)
}
